### PR TITLE
gxml: init at 0.20.3

### LINKED ISSUE #194803: Add GXml package
**Project description**

> GXml provides a GObject API for manipulating XML and a Serializable
framework from GObject to XML.

From [`README.md`](https://gitlab.gnome.org/GNOME/gxml/-/blob/master/README)

This is essentially an XML reader/writer library based on libxml2 written in Vala and uses GObjects.

**Metadata**

* homepage URL: https://gitlab.gnome.org/GNOME/gxml
* source URL: https://gitlab.gnome.org/GNOME/gxml
* license: lgpl2.1
* platforms: unix, linux, darwin, ...

### PR DESCRIPTION
**Pull request by jmarmstrong1207**
## Description of changes

Init gxml. It provides a GObject API for manipulating XML and a Serializable framework from GObject to XML. It's required for packages such as Planify, where their newest update v4.5 requires gxml. the derivation file has been sourced from the flake.nix in the gxml repository [linked here](https://gitlab.gnome.org/GNOME/gxml/-/blob/master/flake.nix?ref_type=heads)

Closes #194803 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by jmarmstrong1207** [comment]
Currently, there are a few issues I've found and not sure how to fix:
- cpp files that try accessing it via #include <gxml/gxml.h> won't find it because it's being installed at `/include/gxml-0.20/gxml/gxml.h` instead of `/include/gxml/gxml.h`. I'm not sure of the standard way of fixing this.
- When the above is fixed, the cpp file has trouble doing `#include <glib-object.h>`. adding glib to buildInputs does not work.

**Comment by jtojnar** [comment]
> Currently, there are a few issues I've found and not sure how to fix:
> 
>     * cpp files that try accessing it via #include <gxml/gxml.h> won't find it because it's being installed at `/include/gxml-0.20/gxml/gxml.h` instead of `/include/gxml/gxml.h`. I'm not sure of the standard way of fixing this.
> 
>     * When the above is fixed, the cpp file has trouble doing `#include <glib-object.h>`. adding glib to buildInputs does not work.
> 
> 
> I made the package.nix based off of the flake.nix as I've said before, so I'm not sure why these issues are happening. This is my first PR afterall!

How are you building the files? Using pkg-config to get the cflags is the recommended method.

File: pkgs/by-name/gx/gxml/package.nix [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    domain = "gitlab.gnome.org";
+    owner = "GNOME";
+    repo = "gxml";
+    rev = version;
+    hash = "sha256-GlctGxsLyQ2kPV3oBmusRiouG4PPncBTh3vgxhVaQOo=";
+  };
+
+  outputs = [ "out" "dev" "devdoc" ];
+
+  PKG_CONFIG_GOBJECT_INTROSPECTION_1_0_GIRDIR = "${placeholder "dev"}/share/gir-1.0";
```
Comment (jtojnar) [inline pkgs/by-name/gx/gxml/package.nix:27]: Are these really necessary? If yes, they should go to env attribute to work once we switch to __structuredAttrs.

File: pkgs/by-name/gx/gxml/package.nix [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  PKG_CONFIG_GOBJECT_INTROSPECTION_1_0_TYPELIBDIR = "${placeholder "out"}/lib/girepository-1.0";
+
+  doCheck = true;
+  enableParallelBuilding = true;
```
Comment (jtojnar) [inline pkgs/by-name/gx/gxml/package.nix:31]: Meson builds in parallel by default.

File: pkgs/by-name/gx/gxml/package.nix [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    gobject-introspection
+  ];
+
+  buildInputs = [
+    libxml2
+    glib
+    libgee
+  ];
+
+  propagatedBuildInputs = buildInputs;
```
Comment (jtojnar) [inline pkgs/by-name/gx/gxml/package.nix:47]: You do not need to duplicate the `buildInputs`, `propagatedBuildInputs` already set `buildInputs`.

**Comment by jmarmstrong1207** [comment]
> > Currently, there are a few issues I've found and not sure how to fix:
> > ```
> > * cpp files that try accessing it via #include <gxml/gxml.h> won't find it because it's being installed at `/include/gxml-0.20/gxml/gxml.h` instead of `/include/gxml/gxml.h`. I'm not sure of the standard way of fixing this.
> > 
> > * When the above is fixed, the cpp file has trouble doing `#include <glib-object.h>`. adding glib to buildInputs does not work.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I made the package.nix based off of the flake.nix as I've said before, so I'm not sure why these issues are happening. This is my first PR afterall!
> 
> How are you building the files? Using pkg-config to get the cflags is the recommended method.

this is my mkShell. I have added pkg-config per your suggestion and it cannot find glib-object.h still
```
{ pkgs ? import <nixpkgs> {} }:
  let
    gxml = import /home/critzlez/.config/nixos/gxml.nix;
  in
  with pkgs;
  pkgs.mkShell {
    buildInputs = [
      pkg-config
      gcc
      glib
      gxml
    ];
}
```

If I solve the error where it can't find <gxml/gxml.h> by moving the gxml folder, this is the exact error:
`fatal error: glib-object.h: No such file or directory` 

gxml.h itself references #include <glib-object.h>, which is where it's coming from

The cpp file is simply this:
```
#include <gxml/gxml.h>

int main()
{
	return 0;
}
```

**Comment by jmarmstrong1207** [comment]
Actually, that problem might have been because of incorrect compiler settings when I was compiling the test cpp file. It compiles perfectly when I try compiling Planify v4.5 which requires Gxml!

I'll go ahead and add changes based on your reviews. First, I need to know these:
- What is the correct way to set the /include path? As I have said before, Planify v4.5 won't compile because it couldnt find <gxml/gxml.h>. Manually moving it to /include/gxml/gxml.h from /include/gxml-0.20/gxml/gxml.h fixes it, but doesn't seem like the correct way
- Could you explain further with your review about env attributes and __structuredAttrs? And which lines specifically are you talking about

File: pkgs/by-name/gx/gxml/package.nix [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+  src = fetchFromGitLab {
+    domain = "gitlab.gnome.org";
+    owner = "GNOME";
+    repo = "gxml";
+    rev = version;
+    hash = "sha256-GlctGxsLyQ2kPV3oBmusRiouG4PPncBTh3vgxhVaQOo=";
+  };
+
+  outputs = [ "out" "dev" "devdoc" ];
```
Comment (jtojnar) [inline pkgs/by-name/gx/gxml/package.nix:25]: GNOME packages prefer having the `outputs` attribute between `version` and `src` see https://discourse.nixos.org/t/document-attribute-ordering-in-package-expressions/4887 and https://github.com/jtojnar/nixpkgs-hammering/blob/main/explanations/attribute-ordering.md.

File: pkgs/by-name/gx/gxml/package.nix [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+, meson
+, ninja
+, pkg-config
+, stdenv
+, lib
+, libgee
+, libxml2
+, vala
+}:
+stdenv.mkDerivation rec {
```
Comment (jtojnar) [inline pkgs/by-name/gx/gxml/package.nix:13]: Would be nice to replace the `rec` with [`finalAttrs` pattern](https://nixos.org/manual/nixpkgs/stable/#mkderivation-recursive-attributes).

**Comment by jmarmstrong1207** [comment]
> cc foo.c $(pkg-config gxml-0.20 --cflags --libs)

I got it to find gxml. Planify already uses meson but for some reason it was adding gxml-0.20/include in the path folder when it should've been doing gxml-0.20-dev/include. I fixed it by removing "dev" in the outputs array. Is this fine?

**Comment by jmarmstrong1207** [comment]
Resolved your review comments, and pointed the git repo to the PR that fixes the issue for NixOS.

**Comment by jmarmstrong1207** [comment]
added prebuildphase fixing the include directory file. Should be temporary until the PR in gxml fixes it. Not sure if git push --force is preferred to have clean commits, but i've done so

File: pkgs/by-name/gx/gxml/package.nix [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  ];
+
+  propagatedBuildInputs = [
+    libxml2
+    glib
+    libgee
+  ];
+
+  postPatch = ''
+    substituteInPlace gxml/gxml.pc.in --replace-fail "@prefix@/include" "@includedir@"
```
Comment (bobby285271) [inline pkgs/by-name/gx/gxml/package.nix:44]: I doubt if this actually works (did you test building planify 4.5?), I don't see `includedir` being set in [configuration](https://gitlab.gnome.org/GNOME/gxml/-/blob/0.20.3/gxml/meson.build?ref_type=tags#L6) :

On 4b35466df18438d24b3e6a23cb783170a352975c:

```
$ cat $(nix-build -A gxml.dev)/lib/pkgconfig/gxml-0.20.pc | grep includedir
includedir=
Cflags: -I${includedir}/gxml-0.20
```

I pushed some small fixups for this, I think replacing this with `${placeholder "dev"}/include` should be good

**Review by jtojnar — APPROVED** [review]

**Pull request merged**